Update ECAL and HCAL reconstruction to run on multple GPUs

vkhristenko opened the pull request:
#### PR description:

superseeds #498 

hcal and ecal are done together in here cause of the change `CUDADataFormats/HcalCommon` -> `CUDADataFormats/CaloCommon` which both hcal and ecal now depend on. this avoids duplication...

this is to allow hcal and ecal running on a node with multiple gpus.
all the modules have been updated for that and now basically no protection for cuda service is needed.
__note: Ecal RecHit was only updated but not validated (fillDescriptions is not self-sufficient)__ @amassiro  

the only thing in this pr is that the newly added conditions' Records should be moved to CondFormats/DataRecord eventually.

#### PR validation:

using standalone execs

fwyzard commented on RecoLocalCalo/EcalRecProducers/plugins/AmplitudeComputationCommonKernels.cu:
```diff
@@ -247,8 +263,8 @@ namespace ecal {
-            //energies[ch] = amplitudes[ch][sample_max];
-            energies[ch] = amplitude;
+            //energies[inputCh] = amplitudes[ch][sample_max];
+            energies[inputCh] = amplitude;
```
why the change from `ch` to `inputCh` ?
what is the difference ?

vkhristenko replied:
previously there was a single ptr for the output collections (eb was first and then ee). now i switched to 2 ptrs, one for each collection given that each collection has its own unique_ptr that manages memory. `inputCh` is computed having this in mind... ch is just a linear idx.

fwyzard replied:
understood, thanks

fwyzard commented on RecoLocalCalo/EcalRecProducers/plugins/AmplitudeComputationCommonKernels.cu:
```diff
@@ -65,6 +71,16 @@ namespace ecal {
+      ARRANGE(g_pedestal);
+      ARRANGE(dids_out);
+      ARRANGE(flags);
+
```
could you `#undef ARRANGE` once you are done using it ?

fwyzard commented on RecoLocalCalo/EcalRecProducers/plugins/AmplitudeComputationKernels.cu:
```diff
@@ -348,12 +358,12 @@ namespace ecal {
-        chi2s[idx] = chi2;
-        energies[idx] = resultAmplitudes(5);
+        chi2s[inputCh] = chi2;
+        energies[inputCh] = resultAmplitudes(5);
```
same question here: what is the difference between `idx` and `inputCh` ?

vkhristenko replied:
same as above

fwyzard commented on EventFilter/EcalRawToDigi/plugins/EcalRawToDigiGPU.cc:
```diff
@@ -97,6 +85,40 @@ void EcalRawToDigiGPU::acquire(edm::Event const& event,
+
+  // input cpu data
+  ecal::raw::InputDataCPU inputCPU = {
+    cms::cuda::make_host_unique<unsigned char[]>(
```
comment for the future (after CMSSW 11.1.0 / CUDA 11.0 / c++17): would it make sense to use `std::byte` instead of `unsigned char` ?

vkhristenko replied:
I would say we should be consistent with what sits inside of the FEDRawData... although we do make a copy before sending to the device...

fwyzard commented on RecoLocalCalo/HcalRecProducers/src/DeclsForKernels.h:
```diff
@@ -75,26 +82,28 @@ namespace hcal {
+          pulseMatrices, pulseMatricesM, pulseMatricesP;
+      cms::cuda::device::unique_ptr<int8_t[]> soiSamples;
+
+      /*
```
Could you delete these, or there is some use in keeping them ?

vkhristenko replied:
addressed

fwyzard replied:
Thanks

fwyzard commented:
Sorry for taking long to go through this, my plan is to
  - finish review the changes
  - check that it works fine with 2 GPUs

I am also using this PR as an excuse to add the HCAL-only workflows to the validation, but if that takes too long I'll just go ahead and merge it.

vkhristenko commented:
@fwyzard np. i tested this on cmg-gpu1080...

fwyzard commented:
I haven't gotten to the HCAL workflow yet - but this PR already breaks the ECAL-only workflow, and the HLT customisation.

Looks like the reasons are:
  - the old parameters left in the HLT workflow
  - the new ESProdcers missing from the HLT and offline workflows

I can try to add them ...

fwyzard commented:
What is the configuration for `EcalRecHitParametersGPUESProducer` supposed to look like ?

The autogenerated cfi is
```python
import FWCore.ParameterSet.Config as cms

ecalRecHitParametersGPUESProducer = cms.ESSource('EcalRecHitParametersGPUESProducer',
  ChannelStatusToBeExcluded = cms.VPSet(
    cms.PSet()
  ),
  appendToDataLabel = cms.string('')
)
```
but using it fails with

> ----- Begin Fatal Exception 08-Jul-2020 19:06:29 CEST-----------------------
> An exception of category 'Configuration' occurred while
>    [0] Constructing the EventProcessor
>    [1] Validating configuration of ESProducer or ESSource of type EcalRecHitParametersGPUESProducer with label: 'ecalRecHitParametersGPUESProducer'
> Exception Message:
> Missing required parameter.  It should have label "kDAC" and have type "tracked string".
> The description has no default.  The parameter must be defined in the configuration
> ----- End Fatal Exception -------------------------------------------------

vkhristenko commented:
The auto generated cfi is not self sufficient...I indicated this in pr
description. Andrea should provide input here.

On Wed, 8 Jul 2020 at 19:17, Andrea Bocci <notifications@github.com> wrote:

> What is the configuration for EcalRecHitParametersGPUESProducer supposed
> to look like ?
>
> The autogenerated cfi is
>
> import FWCore.ParameterSet.Config as cms
> ecalRecHitParametersGPUESProducer = cms.ESSource('EcalRecHitParametersGPUESProducer',
>   ChannelStatusToBeExcluded = cms.VPSet(
>     cms.PSet()
>   ),
>   appendToDataLabel = cms.string('')
> )
>
> but using it fails with
>
> ----- Begin Fatal Exception 08-Jul-2020 19:06:29
> CEST-----------------------
> An exception of category 'Configuration' occurred while
> [0] Constructing the EventProcessor
> [1] Validating configuration of ESProducer or ESSource of type
> EcalRecHitParametersGPUESProducer with label:
> 'ecalRecHitParametersGPUESProducer'
> Exception Message:
> Missing required parameter. It should have label "kDAC" and have type
> "tracked string".
> The description has no default. The parameter must be defined in the
> configuration
> ----- End Fatal Exception -------------------------------------------------
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-patatrack/cmssw/pull/502#issuecomment-655648845>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABSFUCNTUGY7O46ALF5ZULLR2SS2ZANCNFSM4OUMZMDA>
> .
>

fwyzard commented:
I see - I hadn't understood that the comment was about the EDProducer/ESProducer split, thanks for the clarification.

vkhristenko commented:
> I see - I hadn't understood that the comment was about the EDProducer/ESProducer split, thanks for the clarification.

ok, I think that you need to use this guy https://github.com/cms-patatrack/cmssw/blob/CMSSW_11_1_X_Patatrack/RecoLocalCalo/EcalRecProducers/python/ecalRecHitGPU_cfi.py

but again the EDProducer itself does not really have self-sufficient `fillDescriptions`. And when i moved to ES..., I moved only what was present in the EDProducer. Therefore, I think somewhere in the customizations there should be loading of this cfi... @amassiro

amassiro commented:
> > I see - I hadn't understood that the comment was about the EDProducer/ESProducer split, thanks for the clarification.
> 
> ok, I think that you need to use this guy https://github.com/cms-patatrack/cmssw/blob/CMSSW_11_1_X_Patatrack/RecoLocalCalo/EcalRecProducers/python/ecalRecHitGPU_cfi.py
> 
> but again the EDProducer itself does not really have self-sufficient `fillDescriptions`. And when i moved to ES..., I moved only what was present in the EDProducer. Therefore, I think somewhere in the customizations there should be loading of this cfi... @amassiro




Thanks @vkhristenko !

I think that with this 
https://github.com/cms-patatrack/cmssw/pull/504
the configuration producer should be ok.

I tested with the usual test configuration under RecoLocalCalo/EcalRecProducers/test/testEcalRechitProducer_cfg.py

fwyzard commented:
Validated together with #504, #507 and #508.
